confgenerator/confmerger.go

```diff
@@ -0,0 +1,149 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package confgenerator provides functions to generate subagents configuration from unified agent.
+package confgenerator
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/viper"
+)
+
+var (
+	builtInConf = map[string]string{
+		"linux": `logging:
+  receivers:
+    syslog:
+      type: files
+      include_paths:
+      - /var/log/messages
+      - /var/log/syslog
+  exporters:
+    google:
+      type: google_cloud_logging
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [syslog]
+        exporters: [google]
+
+metrics:
+  receivers:
+    hostmetrics:
+      type: hostmetrics
+      collection_interval: 60s
+  exporters:
+    google:
+      type: google_cloud_monitoring
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [hostmetrics]
+        exporters: [google]`,
+
+		"windows": `logging:
+  receivers:
+    windows_event_log:
+      type: windows_event_log
+      channels: [System,Application,Security]
+  exporters:
+    google:
+      type: google_cloud_logging
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [windows_event_log]
+        exporters: [google]
+
+metrics:
+  receivers:
+    hostmetrics:
+      type: hostmetrics
+      collection_interval: 60s
+    mssql:
+      type: mssql
+      collection_interval: 60s
+    iis:
+      type: iis
+      collection_interval: 60s
+  exporters:
+    google:
+      type: google_cloud_monitoring
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [hostmetrics,mssql,iis]
+        exporters: [google]`,
+	}
+)
+
+func MergeConfFiles(builtInConfPath, userConfPath, mergedConfPath, platform string) error {
+	builtInConfFolder := path.Dir(builtInConfPath)
+	builtInConfFullName := path.Base(builtInConfPath)
+	builtInConfName := strings.TrimSuffix(builtInConfFullName, filepath.Ext(builtInConfFullName))
+	userConfFolder := path.Dir(userConfPath)
+	userConfFullName := path.Base(userConfPath)
+	userConfName := strings.TrimSuffix(userConfFullName, filepath.Ext(userConfFullName))
+	mergedConfFolder := path.Dir(mergedConfPath)
+
+	// Write the built-in conf to disk for debugging purpose.
+	f, err := os.Create(builtInConfPath)
+	if err != nil {
+		return fmt.Errorf("error creating the built-in config file %q: %w \n", builtInConfPath, err)
+	}
+	defer f.Close()
+	if _, err = f.WriteString(builtInConf[platform]); err != nil {
+		return fmt.Errorf("error writing the built-in config file %q: %w \n", builtInConfPath, err)
+	}
+
+	// Init Viper to process YAML format configs.
+	viper := viper.New()
```

Comment:
Viper is a huge monstrosity that is overkill for Ops Agent. Using viper to parse/merge these config files imports a ton of really complicated logic. Can we really not just do something far simpler and easier to understand for GA? (i.e. "read each config file separately, and merge the top level maps")

Reply:
One of the goals is to avoid inventing a new language that increased maintenance burden. Viper meets the use cases we needed and is used by Datadog and Otel. It's more of an implementation detail though. I'm interested in the `a ton of really complicated logic` though. Could you elaborate a bit more? (We could also take it offline or to internal docs)

Reply:
I don't understand what you mean by "inventing a new language". We already have yaml, and there's no need to change that. This could be as simple as (pseudocode):

```
var config Config
for _, f := range file {
  fc := load(f)
  for k, v := range fc.Metrics.Receivers {
    config.Metrics.Receivers[k] = v
  }
  // Repeat for the other 5 fieldss
}
```

where `load` is just the existing code we have for reading the YAML file.

Reply:
Replaced Viper with some custom code.